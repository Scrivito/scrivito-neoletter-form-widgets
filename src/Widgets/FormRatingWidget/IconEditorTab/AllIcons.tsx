import * as React from "react";
import { bootstrapIcons } from "../../FormStepContainerWidget/utils/bootstrapIcons";
interface AllIconsProps {
  setWidgetIcon: (event: React.BaseSyntheticEvent, icon: string) => void;
  currentIcon: string;
  hide: boolean;
}
interface IconsProps {
  icons: string[];
  currentIcon: string;
  setWidgetIcon: (event: React.BaseSyntheticEvent, icon: string) => void;
}

export const AllIcons: React.FC<AllIconsProps> = ({
  setWidgetIcon,
  currentIcon,
  hide
}) => {
  const icons = React.useMemo(() => {
    return bootstrapIcons;
  }, []);

  if (hide) return null;

  return (
    <div id="icons">
      {
        <Icons
          icons={icons}
          currentIcon={currentIcon}
          setWidgetIcon={setWidgetIcon}
        />
      }
    </div>
  );
};

const Icons: React.FC<IconsProps> = ({ icons, currentIcon, setWidgetIcon }) => {
  return (
    <div className="row">
      {icons.map((icon, innerIndex) => {
        const cssIcon = `bi-${icon}`;
        const aClassNames: string[] = [];
        if (currentIcon === cssIcon) aClassNames.push("active");

        return (
          <div key={`${icon}-${innerIndex}`} className="icon-container">
            <a
              href="#"
              className={aClassNames.join(" ")}
              onClick={e => setWidgetIcon(e, cssIcon)}>
              <i className={["bs-icon", cssIcon].join(" ")} aria-hidden="true" />
              <span className="sr-only">Example of </span>
              {icon}
            </a>
          </div>
        );
      })}
    </div>
  );
};
