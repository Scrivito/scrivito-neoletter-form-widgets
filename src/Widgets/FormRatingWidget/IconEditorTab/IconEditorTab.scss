@mixin srOnly() {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  border: 0;
}
@mixin icons() {
  --bs-gutter-x: 1.5rem;
  --bs-gutter-y: 0;
  display: flex;
  flex-wrap: wrap;
  margin-top: calc(-1 * var(--bs-gutter-y));
  margin-right: calc(-0.5 * var(--bs-gutter-x));
  margin-left: calc(-0.5 * var(--bs-gutter-x));
}
@mixin activeIcon() {
  background-color: #fff;
  color: #333;
  text-decoration: none;
  box-shadow: 0 0 0 2px #ff6b6b inset;
}
.neoletter-form-icon-editor-tab {
  .icon-editor-preview {
    background: rgba(255, 255, 255, 0.8);
    padding: 14px 10px 10px 10px;
    border-radius: 4px;
    margin: 0 0 10px 0;
    display: block;
    width: 60px;
    text-align: center;
    line-height: 1.3em;
    .bi {
      font-size: 26px;
    }
  }
  .bi-icon {
    width: 25%;
    a {
      display: flex;
      align-items: center;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      text-decoration: none;
      color: #222;
      line-height: 33px;
      padding: 2px 10px;
      border-radius: 4px;
      font-size: 13px;
      margin: 2px 0;
      .sr-only {
        @include srOnly();
      }
      .bi {
        margin-right: 10px;
        color: var(--bs-primary, #5c9dcd);
        font: normal normal normal 14px "bootstrap-icons";
      }
      &:hover {
        background-color: #fff;
        color: #333;
        text-decoration: none;
        .bi {
          font-size: 23px;
        }
        .text-muted {
          color: #bbe2d5;
        }
      }
      &.active {
        @include activeIcon();
        .bi {
          font-size: 24px;
        }
        .text-muted {
          color: #bbe2d5;
        }
      }
    }
  }
  .row {
    @include icons();
  }
}
#neoletter-icon-search {
  position: relative;
  display: flex;
  align-items: center;
  .bi {
    color: #555;
  }
  label {
    position: absolute;
    left: 17px;
  }
  #search-input {
    flex-grow: 1;
    padding: 1px 5px 1px 43px !important;
    border-radius: 23px;
    font-size: 13px;
    line-height: 25px;
    &:focus {
      box-shadow: none;
    }
  }
  #search-clear.bi {
    text-decoration: none;
    position: absolute;
    right: 10px;
    color: #777;
    &:hover {
      color: #8f8f8f;
    }
  }
  .sr-only {
    @include srOnly();
  }
  #search-results {
    .row {
      @include icons();
    }
  }
}

// -- dark mode -- //

.scrivito_dark {
  .neoletter-form-icon-editor-tab .icon-editor-preview {
    background: #5a5a5a;
  }
  .neoletter-form-icon-editor-tab .headline,
  .neoletter-form-icon-editor-tab .icon-editor-preview .bi {
    color: #eee;
  }
  #neoletter-icon-search .bi {
    color: #aaa;
  }
  #neoletter-icon-search #search-input {
    background: #333;
    border: 1px solid #333;
    color: #777;
  }
  .neoletter-form-icon-editor-tab .bi-icon a {
    color: #eee;
    &:hover,
    &:active {
      background: #eee;
      color: black;
      text-decoration: none;
      box-shadow:
        0 0 3px rgba(0, 0, 0, 0.2),
        0 0 0 2px #426698 inset;
    }
    &.active {
      @include activeIcon();
    }
  }
}
