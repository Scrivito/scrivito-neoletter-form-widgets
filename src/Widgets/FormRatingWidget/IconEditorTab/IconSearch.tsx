import * as React from "react";

interface IconSearchProps {
  setSearchValue: React.Dispatch<React.SetStateAction<string>>;
  searchValue: string;
}
interface ClearSearchButtonProps {
  setSearchValue: React.Dispatch<React.SetStateAction<string>>;
  searchValue: string;
}

export const IconSearch: React.FC<IconSearchProps> = ({
  setSearchValue,
  searchValue
}) => {
  return (
    <div id="neoletter-icon-search">
      <label htmlFor="search-input">
        <i className="bs-icon bi-search" aria-hidden="true" />
        <span className="sr-only">Search icons</span>
      </label>
      <input
        id="search-input"
        className="form-control"
        placeholder="Search icons"
        autoComplete="off"
        spellCheck="false"
        autoCorrect="off"
        tabIndex={1}
        value={searchValue}
        onChange={event => {
          event.preventDefault();
          event.stopPropagation();

          setSearchValue(event.target.value);
        }}
      />
      <ClearSearchButton
        setSearchValue={setSearchValue}
        searchValue={searchValue}
      />
    </div>
  );
};

const ClearSearchButton: React.FC<ClearSearchButtonProps> = ({
  setSearchValue,
  searchValue
}) => {
  if (!searchValue.length) return null;

  return (
    <a
      id="search-clear"
      href="#"
      className="bs-icon bi-x-circle"
      aria-hidden="true"
      onClick={() => setSearchValue("")}>
      <span className="sr-only">Clear search</span>
    </a>
  );
};
