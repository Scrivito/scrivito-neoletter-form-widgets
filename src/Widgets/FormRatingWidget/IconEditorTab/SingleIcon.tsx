import * as React from "react";

interface SingleIconProps {
  icon: string;
  setWidgetIcon: (event: React.BaseSyntheticEvent, icon: string) => void;
  currentIcon: string;
}
export const SingleIcon: React.FC<SingleIconProps> = ({
  icon,
  setWidgetIcon,
  currentIcon
}) => {
  const cssIcon = `bi-${icon}`;

  const aClassNames: string[] = [];
  if (currentIcon === cssIcon) aClassNames.push("active");

  return (
    <div className="bi-icon">
      <a
        href="#"
        className={aClassNames.join(" ")}
        onClick={e => setWidgetIcon(e, cssIcon)}>
        <i className={`bi ${cssIcon}`}></i>
        <span className="sr-only">Example of </span>
        {icon}
      </a>
    </div>
  );
};
