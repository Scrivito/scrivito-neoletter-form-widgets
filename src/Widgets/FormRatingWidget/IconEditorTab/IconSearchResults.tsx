import * as React from "react";
import Fuse from "fuse.js";
import { bootstrapIcons } from "../../FormStepContainerWidget/utils/bootstrapIcons";
import { SingleIcon } from "./SingleIcon";

interface IconSearchResultsProps {
  searchValue: string;
  setWidgetIcon: (event: React.BaseSyntheticEvent, icon: string) => void;
  currentIcon: string;
}
export const IconSearchResults: React.FC<IconSearchResultsProps> = ({
  searchValue,
  setWidgetIcon,
  currentIcon
}) => {
  const fuse = React.useMemo(() => {
    const fuseOptions = {
      threshold: 0.2
    };
    return new Fuse(bootstrapIcons, fuseOptions);
  }, []);

  if (!searchValue.length) return null;

  const results = fuse.search(searchValue);

  return (
    <div id="search-results">
      <div className="scrivito_detail_label">
        <span>{`Search for '${searchValue}'`}</span>
      </div>
      <div className="row">
        {results.map((result, index) => {
          const icon = result.item;
          return (
            <SingleIcon
              key={`${icon}${index}`}
              icon={icon}
              currentIcon={currentIcon}
              setWidgetIcon={setWidgetIcon}
            />
          );
        })}
      </div>
    </div>
  );
};
