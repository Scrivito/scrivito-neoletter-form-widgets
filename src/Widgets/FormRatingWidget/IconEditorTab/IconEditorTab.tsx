import * as React from "react";
import * as Scrivito from "scrivito";
import { isEmpty } from "lodash";
import { AllIcons } from "./AllIcons";
import { IconSearch } from "./IconSearch";
import { IconSearchResults } from "./IconSearchResults";
import { getIconColor } from "../../FormStepContainerWidget/utils/getIconColor";
import "./IconEditorTab.scss";

interface IconEditorTabProps {
  widget: Scrivito.Widget;
}

export const IconEditorTab: React.FC<IconEditorTabProps> = ({ widget }) => {
  const [searchValue, setSearchValue] = React.useState("");
  const currentIcon = widget.get("icon") as string;
  const color = getIconColor(widget);
  const uiContext = Scrivito.uiContext();
  if (!uiContext) return null;

  return (
    <div className={`scrivito_${uiContext.theme}`}>
      <div className="neoletter-form-icon-editor-tab">
        <div className="scrivito_detail_content">
          <div className="scrivito_detail_label">
            <span>Preview</span>
          </div>
          <div className="icon-editor-preview">
            <i
              className={`bi ${currentIcon}`}
              style={{
                color: color
              }}></i>
          </div>

          {Scrivito.canWrite() && (
            <>
              <IconSearch
                searchValue={searchValue}
                setSearchValue={newSearchValue => {
                  if (searchValue !== newSearchValue) {
                    setSearchValue(newSearchValue);
                  }
                }}
              />
              <IconSearchResults
                currentIcon={currentIcon}
                searchValue={searchValue}
                setWidgetIcon={setWidgetIcon}
              />
              <AllIcons
                currentIcon={currentIcon}
                hide={!isEmpty(searchValue)}
                setWidgetIcon={setWidgetIcon}
              />
            </>
          )}
        </div>
      </div>
    </div>
  );

  function setWidgetIcon(event: React.BaseSyntheticEvent, icon: string): void {
    event.preventDefault();
    event.stopPropagation();
    widget.update({ icon });
  }
};
