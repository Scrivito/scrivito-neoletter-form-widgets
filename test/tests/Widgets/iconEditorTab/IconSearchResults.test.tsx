import React from "react";
import { render } from "@testing-library/react";
import { IconSearchResults } from "../../../../src/Widgets/FormRatingWidget/IconEditorTab/IconSearchResults";
import renderer from "react-test-renderer";

const mockSetWidgetIcon = jest.fn();
const searchValue = "moon";
const currentIcon = "bi-current";

describe("IconSearchResults component", () => {
  it("renders search results correctly for the searchValue", () => {
    const { getByText } = render(
      <IconSearchResults
        searchValue={searchValue}
        setWidgetIcon={mockSetWidgetIcon}
        currentIcon={currentIcon}
      />
    );

    const searchLabel = getByText(`Search for '${searchValue}'`);
    expect(searchLabel).toBeInTheDocument();

    const expectedIcons = [
      "moon-fill",
      "moon-stars",
      "moon-stars-fill",
      "cloud-moon",
      "cloud-moon-fill"
    ];

    expectedIcons.forEach(icon => {
      const iconText = getByText(icon);
      expect(iconText).toBeInTheDocument();
    });
  });

  it("does not render search results when searchValue is empty", () => {
    const { container } = render(
      <IconSearchResults
        searchValue=""
        setWidgetIcon={mockSetWidgetIcon}
        currentIcon=""
      />
    );

    const searchResults = container.querySelector("#search-results");
    expect(searchResults).toBeNull();
  });

  it("renders correctly", () => {
    const searchValue = "moon";

    const tree = renderer
      .create(
        <IconSearchResults
          searchValue={searchValue}
          setWidgetIcon={mockSetWidgetIcon}
          currentIcon={currentIcon}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
