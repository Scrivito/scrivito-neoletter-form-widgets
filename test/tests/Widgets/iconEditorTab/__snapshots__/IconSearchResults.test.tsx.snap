// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconSearchResults component renders correctly 1`] = `
<div
  id="search-results"
>
  <div
    className="scrivito_detail_label"
  >
    <span>
      Search for 'moon'
    </span>
  </div>
  <div
    className="row"
  >
    <div
      className="bi-icon"
    >
      <a
        className=""
        href="#"
        onClick={[Function]}
      >
        <i
          className="bi bi-moon"
        />
        <span
          className="sr-only"
        >
          Example of 
        </span>
        moon
      </a>
    </div>
    <div
      className="bi-icon"
    >
      <a
        className=""
        href="#"
        onClick={[Function]}
      >
        <i
          className="bi bi-moon-fill"
        />
        <span
          className="sr-only"
        >
          Example of 
        </span>
        moon-fill
      </a>
    </div>
    <div
      className="bi-icon"
    >
      <a
        className=""
        href="#"
        onClick={[Function]}
      >
        <i
          className="bi bi-moon-stars"
        />
        <span
          className="sr-only"
        >
          Example of 
        </span>
        moon-stars
      </a>
    </div>
    <div
      className="bi-icon"
    >
      <a
        className=""
        href="#"
        onClick={[Function]}
      >
        <i
          className="bi bi-moon-stars-fill"
        />
        <span
          className="sr-only"
        >
          Example of 
        </span>
        moon-stars-fill
      </a>
    </div>
    <div
      className="bi-icon"
    >
      <a
        className=""
        href="#"
        onClick={[Function]}
      >
        <i
          className="bi bi-cloud-moon"
        />
        <span
          className="sr-only"
        >
          Example of 
        </span>
        cloud-moon
      </a>
    </div>
    <div
      className="bi-icon"
    >
      <a
        className=""
        href="#"
        onClick={[Function]}
      >
        <i
          className="bi bi-cloud-moon-fill"
        />
        <span
          className="sr-only"
        >
          Example of 
        </span>
        cloud-moon-fill
      </a>
    </div>
  </div>
</div>
`;
